plugins/woocommerce/includes/wc-product-functions.php

```diff
@@ -1783,3 +1783,83 @@ function wc_product_attach_featured_image( $attachment_id, $product = null, $sav
 	}
 }
 add_action( 'add_attachment', 'wc_product_attach_featured_image' );
+
+/**
+ * Create an attachment from an external URL.
+ *
+ * @private
+ *
+ * @since 9.4.0
+ * @param string $image_url URL of the image.
+ * @param int    $parent_post_id Optional parent post ID.
+ * @return int Attachment ID.
+ */
+function wc_create_external_media_attachment( $image_url, $parent_post_id = null ) {
+	$file_name = basename( $image_url );
+
+	$args = array(
+		'post_type'      => 'attachment',
+		'post_status'    => 'inherit',
+		'name'           => $file_name,
+		'posts_per_page' => 1,
+	);
+
+	$query      = new WP_Query( $args );
+	$attachment = $query->posts[0] ?? null;
+
+	if ( $attachment ) {
+		return $attachment->ID;
+	}
+
+	$attachment = array(
+		'guid'           => $image_url,
+		'post_mime_type' => 'image/jpeg',
+		'post_title'     => $file_name,
+		'post_content'   => '',
+		'post_status'    => 'inherit',
+	);
+
+	$attach_id = wp_insert_attachment( $attachment, '', $parent_post_id );
+
+	// TODO: Investigate how much resource this consumes.
+	require_once ABSPATH . 'wp-admin/includes/image.php';
+	$attach_data = wp_generate_attachment_metadata( $attach_id, $image_url );
```

Comment:
This process _could_ take some time, we should investigate and if so find a way to skip it (e.g. we can manually generate the meta data since we know what the images are ahead of time)